zkevm-circuits/src/root_circuit/aggregation.rs

```diff
@@ -0,0 +1,698 @@
+                        (instances, accumulators)
+                    })
+                    .collect_vec()
+                    .into_iter()
+                    .unzip::<_, _, Vec<_>, Vec<_>>();
+                let accumulator = {
+                    let as_vk = Default::default();
+                    let as_proof = Vec::new();
+                    let accumulators = accumulators.into_iter().flatten().collect_vec();
+                    let mut transcript =
+                        PoseidonTranscript::new(&loader, Value::known(as_proof.as_slice()));
+                    let proof =
+                        KzgAs::<M>::read_proof(&as_vk, &accumulators, &mut transcript).unwrap();
+                    KzgAs::<M>::verify(&as_vk, &accumulators, &proof).unwrap()
+                };
+
+                let instances = instances
+                    .iter()
+                    .map(|instances| {
+                        instances
+                            .iter()
+                            .map(|instances| {
+                                instances
+                                    .iter()
+                                    .map(|instance| instance.assigned().to_owned())
+                                    .collect_vec()
+                            })
+                            .collect_vec()
+                    })
+                    .collect_vec();
+                let accumulator_limbs = [accumulator.lhs, accumulator.rhs]
+                    .iter()
+                    .map(|ec_point| {
+                        loader
+                            .ecc_chip()
+                            .assign_ec_point_to_limbs(&mut loader.ctx_mut(), ec_point.assigned())
+                    })
+                    .collect::<Result<Vec<_>, Error>>()?
+                    .into_iter()
+                    .flatten()
+                    .collect();
+
+                Ok((instances, accumulator_limbs))
+            },
+        )
+    }
+}
+
+/// Aggregate snarks into a single accumulator and decompose it into
+/// `4 * LIMBS` limbs.
+/// Returns `None` if any given snarks is invalid.
+pub fn aggregate<'a, M: MultiMillerLoop>(
+    params: &ParamsKZG<M>,
+    snarks: impl IntoIterator<Item = Snark<'a, M::G1Affine>>,
+) -> Option<[M::Scalar; 4 * LIMBS]> {
+    let svk = KzgSvk::<M>::new(params.get_g()[0]);
+    let dk = KzgDk::new(svk, params.g2(), params.s_g2());
+
+    let accumulators = snarks
+        .into_iter()
+        .map(|snark| {
+            let mut transcript = PoseidonTranscript::new(snark.proof);
+            let proof = PlonkSuccinctVerifier::<M>::read_proof(
+                &svk,
+                snark.protocol,
+                snark.instances,
+                &mut transcript,
+            )?;
+            PlonkSuccinctVerifier::verify(&svk, snark.protocol, snark.instances, &proof)
+        })
+        .try_collect::<_, Vec<_>, _>()
+        .ok()?
+        .into_iter()
+        .flatten()
+        .collect_vec();
```

Comment:
Perhaps I'm missing something, here I would want to stop earlier if any snark is invalid, I think `try_collect` seems already to be a good choice? Would you like to give an example to show how to adopt `fust` for here 🙏

Reply:
Ohh, I just say it because `fuse` stops once it finds the first error. While `try_collect` executes for all of the iterator elements and later collapses the results into a single error (if any). So you save the computation done for all the snarks after the failing one. But isn't a big deal!

Feel free to ignore!

Reply:
Hmm I don’t think so, `try_collect` also stops immediately once an error  happens (see the doc [here](https://doc.rust-lang.org/src/core/result.rs.html#2031)). So I guess it should be fine.

Reply:
I was under wrong assumptions it seems! Thanks for pointing out!